applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -872,6 +848,29 @@ bool verify_datamode_control(uint16_t time_limit, uint16_t *min_time_limit)
 	return true;
 }
 
+struct at_param_list *slm_get_at_param_list(const char *at_cmd)
+{
+	int err;
+
+	at_params_list_clear(&slm_at_param_list);
+	err = at_parser_params_from_str(at_cmd, NULL, &slm_at_param_list);
+	if (err) {
+		LOG_ERR("AT command parsing failed: %d", err);
+		at_params_list_clear(&slm_at_param_list);
+	}
+
+	return &slm_at_param_list;
+}
+
+void set_default_at_response(void *buf, size_t size)
```

Comment:
I think that I mentioned this to @eivindj-nordic as I was writing the 50th call to this. I kind of like the macro idea, but it would be much nicer, if this did not need to be repeated.

Reply:
I think that it might also be possible to macro the wrappers on the SLM side. Though that would increase the code size as the wrappers would need to be per function.

Reply:
> It would need to be post AT-command callback.

No, it wouldn't I think.
I mean it could work as a post callback if all you want is to fill the response buffer with OK, but I imagine it could be even more magical as a pre callback as it could do both the AT command parsing *and* the filling of the response buffer.

Reply:
Havin it as a pre callback would overwrite the response buffer and the AT-command. I don't think that we can expect that this would be allowed in pre callback. Point was that both are needed.

Reply:
@tomi-font, @SeppoTakalo: how about something like this?

```
#define SLM_AT_CMD_CUSTOM(entry, _filter, _callback)                                               \
	static int _callback(enum at_cmd_type cmd_type, const struct at_param_list *list,          \
			     uint32_t);                                                            \
	static int _callback##_wrapper(char *buf, size_t len, char *at_cmd)                        \
	{                                                                                          \
		const struct at_param_list *list = slm_get_at_param_list(at_cmd);                  \
		const enum at_cmd_type cmd_type = at_parser_cmd_type_get(at_cmd);                  \
		const uint32_t param_count = at_params_valid_count_get(list);                      \
		int ret;                                                                           \
                                                                                                   \
		ret = _callback(cmd_type, list, param_count);                                      \
		if (!ret) {                                                                        \
			return at_cmd_custom_respond(buf, len, "OK\r\n");                          \
		}                                                                                  \
		return ret;                                                                        \
	}                                                                                          \
	AT_CMD_CUSTOM(entry, _filter, _callback##_wrapper);

```

Reply:
It feels a bit like macro abuse.

But not the worst idea. If all the custom CMDs we have, needs there 3 parameters to be parsed, then I guess it is OK.

Reply:
We will look at the proposed changes for the modem library. It should not be a big change to add a return value for a default "OK" response, though the question becomes how minimal we would like to keep the custom commands in the modem library.

Reply:
Not all of them need the values, but many do. A simplified version of this would be a one that only sets the return value.

Reply:
@MarkusLassila can you extract the common parts away from the macro

```c
typedef int slm_at_callback(enum at_cmd_type cmd_type, const struct at_param_list *list,  
			     uint32_t);

/* This implementation is now only once in the flash */
static int slm_cb_wrapper(slm_at_callback *cb, char *buf, size_t len, char *at_cmd){
{
		const struct at_param_list *list = slm_get_at_param_list(at_cmd);
		const enum at_cmd_type cmd_type = at_parser_cmd_type_get(at_cmd);
		const uint32_t param_count = at_params_valid_count_get(list);
		int ret;

		ret = cb(cmd_type, list, param_count);
		if (!ret) {
			return at_cmd_custom_respond(buf, len, "OK\r\n");
		}
		return ret;
}


#define SLM_AT_CMD_CUSTOM(entry, _filter, _callback)                                               \
	static int _callback(enum at_cmd_type cmd_type, const struct at_param_list *list,          \
			     uint32_t);                                                            \
	static int _callback##_wrapper(char *buf, size_t len, char *at_cmd)                        \
	{                                                                                          \
		return slm_cb_wrapper(_callback, buf, len, at_cmd);				   \
	}                                                                                          \
	AT_CMD_CUSTOM(entry, _filter, _callback##_wrapper);

```

Reply:
@MarkusLassila  I like the macro. Though I support @SeppoTakalo's suggestion of having the common wrapper defined only once. But I definitely support going for this.

@eivindj-nordic A return value for a default "OK" response could be nice. And it makes sense that you want to keep the custom commands within certain limits. SLM is quite a particular user of them, and it seems like we can actually get away doing things the way we want in SLM itself.

Reply:
Seems like a good idea. I'll give this a go.

Reply:
Though I suggest changing @SeppoTakalo's

> int slm_cb_wrapper(slm_at_callback *cb, char *buf, size_t len, char *at_cmd)

to

> int slm_cb_wrapper(char *buf, size_t len, char *at_cmd, slm_at_callback *cb)

so that the first 3 received arguments (in ` _callback##_wrapper()`) can be forwarded as-is. No need to re-organize the parameters in the CPU registers. Only the extra callback argument is added. Further reduces the flash usage.

Reply:
Or then you can also have the wrapper be a pre-only callback such that you call `at_cmd_custom_respond()` before the actual AT command handler. This will invalidate `at_cmd` before the actual AT command handler, but you don't use it anymore after having parsed the paramters anyway.
Like that, you wouldn't need to pass the callback to the common wrapper function. You could first call the common wrapper, then the actual callback.
Whichever way.